File: kimdaeyeong/src/programmers/Pg81302.java
package programmers;

import java.util.*;

/**
 * 프로그래머스 81302
 * 거리두기 확인하기
 * lv2
 * https://school.programmers.co.kr/learn/courses/30/lessons/81302
 */
public class Pg81302 {

    static int[] dx = {-1, 0, 1 ,0};
    static int[] dy = {0, -1, 0, 1};
    static boolean[][] visited;
    static int result;

    public static void main(String[] args) {
        String[][] places = {{"POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"}, {"POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"}, {"PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"}, {"OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"}, {"PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"}};

        int[] answer = new int[5];

        for(int k = 0; k < 5; k++) {
            char[][] cur = makePlace(places[k]);
            visited = new boolean[5][5];
            result = 1;

            for(int i = 0; i < 5; i++) {
                for(int j = 0; j < 5; j++) {
                    if(cur[i][j] == 'P' && !visited[i][j]) {
                        visited[i][j] = true;
                        dfs(i, j, i, j, visited, cur);
                    }
                    if(result == 0)
                        break;
                }
                if(result == 0)
                    break;
            }
            answer[k] = result;
        }

        System.out.println(Arrays.toString(answer));
    }

    public static char[][] makePlace(String[] place) {
        char[][] cur = new char[5][5];

        for(int j = 0; j < 5; j++) {
            cur[j] = place[j].toCharArray();
        }

        return cur;
    }

    public static void dfs(int x, int y, int curX, int curY, boolean[][] visited, char[][] cur) {

        for(int i = 0; i < 4; i++) {
            int nextX = curX + dx[i];
            int nextY = curY + dy[i];

            if(nextX < 0 || nextX >= 5 || nextY < 0 || nextY >= 5 || visited[nextX][nextY])
                continue;

            if(cur[nextX][nextY] == 'X')
                continue;

            if(Math.abs(x - nextX) + Math.abs(y - nextY) > 2)
                continue;

            if(cur[nextX][nextY] == 'P') {
                result = 0;
                return;
            }

            visited[nextX][nextY] = true;
            dfs(x, y, nextX, nextY, visited, cur);
            visited[nextX][nextY] = false;
        }
    }
}

File: limsubin/programmers/pg81302.py
# PG 81302: 거리두기 확인하기
# https://school.programmers.co.kr/learn/courses/30/lessons/81302

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def find(p):
    for x in range(5):
        for y in range(5):
            if p[x][y] != 'P':
                continue

            # 응시자가 앉아있는 자리 확인
            for i in range(4):
                mx = x + dx[i]
                my = y + dy[i]

                if mx < 0 or mx >= 5 or my < 0 or my >= 5 or p[mx][my] == 'X':
                    continue

                # 거리두기가 지켜지지 않은 경우
                if p[mx][my] == 'P':
                    return 0

                for j in range(4):
                    mmx = mx + dx[j]
                    mmy = my + dy[j]

                    if mmx < 0 or mmx >= 5 or mmy < 0 or mmy >= 5:
                        continue

                    if mmx == x and mmy == y:
                        continue

                    # 거리두기가 지켜지지 않은 경우
                    if p[mmx][mmy] == 'P':
                        return 0

    return 1

def solution(places):
    answer = []
    for i in range(5):
        answer.append(find(places[i]))
    return answer